### PR TITLE
Refactor data generator

### PR DESCRIPTION
**Pull request by mihaimorariu**
This PR refactors the data generator code. The data generator now outputs ground truth boxes with classes and metadata (shape and scale of rescaled image).

### CODE REVIEW

**Comment by 0x00b1** [comment]
Hi, @mihaimorariu! I appreciate the contribution! I'll review either this evening or tomorrow. Meanwhile, would you mind adding yourself to AUTHORS.rst.

**Comment by 0x00b1** [comment]
@mihaimorariu Would you mind rebasing?

**Comment by mihaimorariu** [comment]
Hi, @0x00b1! Thank you!

I am new to this, but I would gladly like to contribute as well. I have rebased and added myself to AUTHORS.rst, as requested.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 1]
```diff
@@ -1,118 +1,84 @@
...
-
-            if self.batch_index == 0:
-                index_array = numpy.arange(n)
-
-                if shuffle:
-                    index_array = numpy.random.permutation(n)
+import time
 
-            current_index = (self.batch_index * batch_size) % n
+def scale_shape(shape, min_size=224, max_size=224):
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: Nice! We really needed this! 👍 

You should add a test for this function to `tests.preprocessing.object_detection`.

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: It’s a little wonky and unexpected for `shape` to include the channels dimension. I would expect this to work like shape parameters elsewhere (e.g. Keras’ `ImageGenerator`) where it’s a `(width, height)` pair.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 2]
```diff
@@ -1,118 +1,84 @@
...
-            if n > current_index + batch_size:
-                current_batch_size = batch_size
+    # Prevent the biggest axis from being larger than max_size
+    if numpy.round(scale * max_shape) > max_size:
+        scale = max_size / max_shape
 
-                self.batch_index += 1
-            else:
-                current_batch_size = n - current_index
+    return (int(shape[0] * scale), int(shape[1] * scale), shape[2]), scale
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: An implementation that might be clearer:

```Python
    rr, cc = shape

    rr *= scale
    cc *= scale

    return (int(rr), int(cc)), scale
```

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: If you drop the channels dimension, you’ll need to drop the channels dimension here too.

File: keras_rcnn/preprocessing/_object_detection.py (hunk 1 above)
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: It also has the benefit of not needing separate “channels first” and “channels last” branches.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 3]
```diff
@@ -1,118 +1,84 @@
...
 
-            self.total_batches_seen += 1
+def scale_image(image, min_size=224, max_size=224):
+    """
+    Rescales an image according to the heuristics from 'scale_shape'.
+    """
+    target_shape, scale = scale_shape(image.shape, min_size, max_size)
 
-            yield index_array[current_index:current_index + current_batch_size], current_index, current_batch_size
+    return skimage.transform.rescale(image, scale=scale, mode="reflect"), scale
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: I don’t think this function is needed. I’d inline `skimage.transform.rescale`.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 4]
```diff
@@ -1,118 +1,84 @@
...
 
-    def __iter__(self):
-        return self
 
-    def __next__(self, *args, **kwargs):
-        return self.next(*args, **kwargs)
+class DictionaryIterator(keras.preprocessing.image.Iterator):
+    def __init__(self, data, classes, image_data_generator, image_shape, batch_size=1,
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: I’d keep the keyword “dictionary” as it’s more descriptive than `data`.

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: I’d also keep the keyword `generator` since there’re a number of possible generators that could return a `DictionaryGenerator`.

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: You don’t need to include the `image_shape` parameter since many datasets have variable shapes (e.g. PASCAL VOC) and this information is included in the dictionary.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 5]
```diff
@@ -1,118 +1,84 @@
...
+                 shuffle=True, seed=numpy.uint32(time.time() * 1000)):
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: The inner generators should match the semantics and defaults of other Keras generators (like `DirectoryIterator` or `NumPyIterator`) so `shuffle=False`.

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: Likewise, the `seed` semantics should match so it should be `seed=None`.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 6]
```diff
@@ -1,118 +1,84 @@
...
+        self.data                 = data
+        self.classes              = classes
+        self.image_data_generator = image_data_generator
+        self.image_shape          = image_shape
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: `self.image_shape` is unused.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 7]
```diff
@@ -1,118 +1,84 @@
...
+        self.target_shape, _      = scale_shape(image_shape)
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: This shouldn’t be inside the constructor since the `target_shape` is dependent on the image.

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: You’re missing the call to `super()`.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 8]
```diff
@@ -1,118 +1,84 @@
...
-
-        # labels = numpy.expand_dims(labels, 0)
-
-        # metadata = numpy.expand_dims(metadata, 0)
-
-        return [image, boxes, labels, metadata], [boxes, labels]
+            # Create metadata
+            metadata[batch_index, :] = [image.shape[0], image.shape[1], scale]
 
+        return [image_batch, gt_boxes_batch, metadata], None
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: You’re missing the `y` value: `[boxes, labels]`

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: You’re also missing `labels` from `x`.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 9]
```diff
@@ -1,118 +1,84 @@
...
-        pathname = self.dictionary[index]["filename"]
+        # Transformation of images is not under thread lock so it can be done in parallel
+        image_batch    = numpy.zeros((batch_size,) + self.target_shape, dtype=keras.backend.floatx())
+        gt_boxes_batch = numpy.zeros((batch_size, 0, 5),                dtype=keras.backend.floatx())
+        metadata       = numpy.zeros((batch_size, 3),                   dtype=keras.backend.floatx())
 
-        image = skimage.io.imread(pathname)
+        for batch_index, image_index in enumerate(selection):
+            path         = self.data[image_index]["filename"]
+            image        = skimage.io.imread(path, as_grey=(self.target_shape[2] == 1))
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: You don’t need to specify `as_gray`.

**Comment by 0x00b1** [comment]
Thanks, @mihaimorariu! This was great. I requested a few changes. Let me know if you need help!

File: keras_rcnn/preprocessing/_object_detection.py [hunk 10]
```diff
@@ -1,118 +1,84 @@
...
+        return [image_batch, gt_boxes_batch, metadata], None
 
 class ObjectDetectionGenerator:
-    def __init__(self, classes):
-        self.classes = classes
-
-    def flow(self, dictionary, shuffle=True, seed=None):
-        return DictionaryIterator(dictionary, self, shuffle=shuffle, seed=seed)
+    def flow(self, data, classes, image_shape):
+        return DictionaryIterator(data, classes, keras.preprocessing.image.ImageDataGenerator(), image_shape)
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: `keras.preprocessing.image.ImageDataGenerator()` should be `self`

File: keras_rcnn/preprocessing/_object_detection.py [hunk 11]
```diff
@@ -1,118 +1,84 @@
...
 
+        return [image_batch, gt_boxes_batch, metadata], None
 
 class ObjectDetectionGenerator:
-    def __init__(self, classes):
-        self.classes = classes
-
-    def flow(self, dictionary, shuffle=True, seed=None):
-        return DictionaryIterator(dictionary, self, shuffle=shuffle, seed=seed)
+    def flow(self, data, classes, image_shape):
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: I recommend that we keep using `dictionary` rather than `data`. Is `flow_from_dictionary` is clearer?

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: We should drop the `input_shape` parameter.

`shape` should be inferred for situations where images have variable dimensions.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 12]
```diff
@@ -1,118 +1,84 @@
...
+        image_batch    = numpy.zeros((batch_size,) + self.target_shape, dtype=keras.backend.floatx())
+        gt_boxes_batch = numpy.zeros((batch_size, 0, 5),                dtype=keras.backend.floatx())
+        metadata       = numpy.zeros((batch_size, 3),                   dtype=keras.backend.floatx())
 
-        image = skimage.io.imread(pathname)
+        for batch_index, image_index in enumerate(selection):
+            path         = self.data[image_index]["filename"]
+            image        = skimage.io.imread(path, as_grey=(self.target_shape[2] == 1))
+            image, scale = scale_image(image)
+            image        = self.image_data_generator.random_transform(image)
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: I would drop the `random_transform` and `standardize` calls for the moment.

Unfortunately, we can’t easily inherit from `ImageDataGenerator` since we’ll need to write augmentation and pre-processing functions that can manage images and bounding boxes together. We’ll also need to support masks in the near future.

File: keras_rcnn/preprocessing/_object_detection.py [hunk 13]
```diff
@@ -1,118 +1,84 @@
...
 
-        ds = self.dictionary[index]["boxes"]
+            # Set ground truth boxes
+            boxes = self.data[image_index]["boxes"]
+            for i, b in enumerate(boxes):
+                if b["class"] not in self.classes:
+                    raise Exception("Class {} not found in '{}'.".format(b["class"], self.classes))
 
-        boxes = numpy.asarray([[d[k] for k in ['y1', 'x1', 'y2', 'x2']] for d in ds])
+                gt_data = [b["y1"], b["x1"], b["y2"], b["x2"], self.classes[b["class"]]]
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: I believe this should be `['y1', 'x1', 'y2', 'x2']` (i.e. without the class label).

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: We should be yielding (my current understanding):

```Python
# image: (batches, width, height, channels)
# metadata: (batches, 3) where `3` corresponds to the values of `width`, `height`, and `scale`
# bounding boxes: (boxes, 4) where `4` corresponds to `y1`, `x1`, `y2`, and `x2`
# labels: (boxes, classes)

yield (image, metadata, bounding boxes, labels), (bounding boxes, labels)
```

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: (@jhung0 Is my understanding correct?)

File: keras_rcnn/preprocessing/_object_detection.py [hunk 14]
```diff
@@ -1,118 +1,84 @@
...
 
-            yield index_array[current_index:current_index + current_batch_size], current_index, current_batch_size
+    return skimage.transform.rescale(image, scale=scale, mode="reflect"), scale
 
-    def __iter__(self):
-        return self
 
-    def __next__(self, *args, **kwargs):
-        return self.next(*args, **kwargs)
+class DictionaryIterator(keras.preprocessing.image.Iterator):
```
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: Nice. 👍 

I think inheriting from ` keras.preprocessing.image.Iterator` will be easier to maintain than continuing to rely on our implementation.

File: keras_rcnn/preprocessing/_object_detection.py (hunk 13 above)
Comment (jhung0) [inline keras_rcnn/preprocessing/_object_detection.py]: yes, boxes should just have the 4 coordinates. labels has the class info

File: keras_rcnn/preprocessing/_object_detection.py (hunk 1 above)
Comment (mihaimorariu) [inline keras_rcnn/preprocessing/_object_detection.py]: You're right. I have added the unit test.

Comment (mihaimorariu) [inline keras_rcnn/preprocessing/_object_detection.py]: Sure, I can change that. But I think it would be useful to agree on whether we are going to use (width, height) tuples, from now on, or (height, width). I saw different usage in different parts of the code.

File: keras_rcnn/preprocessing/_object_detection.py (hunk 2 above)
Comment (mihaimorariu) [inline keras_rcnn/preprocessing/_object_detection.py]: You're right, this looks cleaner. Fixed.

File: keras_rcnn/preprocessing/_object_detection.py (hunk 4 above)
Comment (mihaimorariu) [inline keras_rcnn/preprocessing/_object_detection.py]: Agreed. I changed that.

File: keras_rcnn/preprocessing/_object_detection.py (hunk 12 above)
Comment (hgaiser) [inline keras_rcnn/preprocessing/_object_detection.py]: Just a note, supporting masks is no issue using the `seed` parameter in `random_transform`. The easiest solution for boxes would be to make a fake image where the bounding box is a square segmentation, augment it using `random_transform` and re-compute the bounding box. Not efficient, but easy.

File: keras_rcnn/preprocessing/_object_detection.py (hunk 13 above)
Comment (hgaiser) [inline keras_rcnn/preprocessing/_object_detection.py]: Ah so we're deviating from py-faster-rcnn here? There, `gt_boxes` is a matrix containing the boxes and labels as a (?, 5) matrix.

Comment (jhung0) [inline keras_rcnn/preprocessing/_object_detection.py]: Yes, the labels here are in one-hot form.

Comment (hgaiser) [inline keras_rcnn/preprocessing/_object_detection.py]: It makes sense to split them, its clearer. Why one-hot form if I may ask?

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: @hgaiser It’s the standard for Keras. It does make some stuff easier (a notable exception is ImageNet’s 21,841 classes 😧).

File: keras_rcnn/preprocessing/_object_detection.py (hunk 11 above)
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: @hgaiser When shape inference works it’d be `(None, None, 3)`. However, until it’s implemented, the model expects fixed dimensions so you’d need to crop or resize.

Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: @hgaiser @mihaimorariu Maybe I misinterpreted the `image_size` parameter? Should it be `target_size`?

File: keras_rcnn/preprocessing/_object_detection.py (hunk 13 above)
Comment (hgaiser) [inline keras_rcnn/preprocessing/_object_detection.py]: Alright then :)

Figured with the sparse versions of the loss functions, for example `sparse_categorical_crossentropy`, it didn't matter much.

File: keras_rcnn/preprocessing/_object_detection.py (hunk 12 above)
Comment (0x00b1) [inline keras_rcnn/preprocessing/_object_detection.py]: @hgaiser 👍

File: keras_rcnn/preprocessing/_object_detection.py (hunk 11 above)
Comment (mihaimorariu) [inline keras_rcnn/preprocessing/_object_detection.py]: `image_shape` is the shape of the original image. `target_shape` is the shape of its rescaled version. So, if I understand correctly, shall we assume that `image_shape` is the same for all images, for now?

Comment (hgaiser) [inline keras_rcnn/preprocessing/_object_detection.py]: @0x00b1 I get that, but generators work with numpy arrays and numpy arrays don't work with shapes like (None, None, 3). A numpy array has to be created, containing any number of images with any kind of shape, how?

If I remember correctly `image_size` is the original image size and this generator would force all images to be of that size (some check for this still needs to be added, would give a clearer error message too; alternative is to reshape all images to this size). In an ideal situation, `image_size` is not forced to one constant size. However, given the problem mentioned above, the most viable solutions is to either force batch size to 1 or force all images to be of the same size. A combination of the two is something I don't see a solution for. I'm very much open to suggestions on this part 👍

File: keras_rcnn/preprocessing/_object_detection.py [hunk 15]
```diff
@@ -1,118 +1,87 @@
...
         with self.lock:
-            index_array, current_index, current_batch_size = next(
-                self.index_generator)
-
-        index = index_array[0]
+            selection, _, batch_size = next(self.index_generator)
 
-        pathname = self.dictionary[index]["filename"]
+        # Labels has num_classes + 1 elements, since 0 is reserved for background.
+        num_classes = len(self.classes.keys())
```
Comment (jhung0) [inline keras_rcnn/preprocessing/_object_detection.py]: could just do len(self.classes)

File: keras_rcnn/preprocessing/_object_detection.py [hunk 16]
```diff
@@ -5,80 +5,83 @@
...
+                label[self.classes[b["class"]]] = 1
+                labels = numpy.append(labels, [[label]], axis = 1)
 
-            # Create metadata
-            metadata[batch_index, :] = [image.shape[0], image.shape[1], scale]
+            # Scale the ground truth boxes to the selected image scale.
+            boxes[batch_index, :, :4] *= self.scale
 
-        return [image_batch, gt_boxes_batch, metadata], None
+        return [images, boxes, labels, self.scale], None
```
Comment (jhung0) [inline keras_rcnn/preprocessing/_object_detection.py]: Why have a None?

Comment (hgaiser) [inline keras_rcnn/preprocessing/_object_detection.py]: Generators must return `X, Y`, but `Y` doesn't make sense in our case because the network generates `Y`. Therefore, `Y` can be ignored.

Comment (jhung0) [inline keras_rcnn/preprocessing/_object_detection.py]: Also, shouldn't metadata have image width/height information too?

Comment (mihaimorariu) [inline keras_rcnn/preprocessing/_object_detection.py]: Do we need to include it, since that information can be obtained from `images`?

File: keras_rcnn/preprocessing/_object_detection.py [hunk 17]
```diff
@@ -37,9 +37,14 @@ def __init__(self, dictionary, classes, generator, batch_size=1, shuffle=False,
         # Compute and store the target image shape.
         cols, rows, channels          = dictionary[0]["shape"]
         self.image_shape              = (rows, cols, channels)
+
         self.target_shape, self.scale = scale_size(self.image_shape[0:2])
```
Comment (jhung0) [inline keras_rcnn/preprocessing/_object_detection.py]: How would the min/max size for scale_size be set?
